File: src/lib/types/crafting.ts
export type RecipeSourceInfo = {
    id: string;
    name: string;
    icon: string;
    recipe_count: number;
};

export type RecipeWithComputed = {
    id: string;
    source_id: string;
    name: string;
    recipe_type: string | null;
    servings: string | null;
    difficulty: string | null;
    time: string | null;
    ingredients: string[];
    steps: string[];
    tags: string[];
    source: string | null;
    image: string | null;
    extra: Record<string, unknown>;
    ingredient_count: number;
    step_count: number;
};

export type RecipeSourceStats = {
    source_id: string;
    source_name: string;
    source_icon: string;
    recipe_count: number;
};

export type RecipeTypeStats = {
    name: string;
    recipe_count: number;
};

export type RecipeStats = {
    total_recipes: number;
    by_source: RecipeSourceStats[];
    by_type: RecipeTypeStats[];
};

export type CraftingData = {
    sources: RecipeSourceInfo[];
    recipes: RecipeWithComputed[];
    stats: RecipeStats;
};

File: src-tauri/src/models/crafting.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// ── Config file ──

#[derive(Deserialize)]
pub struct RecipeSourceFile {
    pub version: u32,
    pub sources: Vec<RecipeSource>,
}

#[derive(Deserialize)]
pub struct RecipeSource {
    pub id: String,
    pub name: String,
    pub path: String,
    pub icon: Option<String>,
}

// ── Response structures ──

#[derive(Serialize)]
pub struct RecipeSourceInfo {
    pub id: String,
    pub name: String,
    pub icon: String,
    pub recipe_count: usize,
}

#[derive(Serialize)]
pub struct RecipeWithComputed {
    pub id: String,
    pub source_id: String,
    pub name: String,
    pub recipe_type: Option<String>,
    pub servings: Option<String>,
    pub difficulty: Option<String>,
    pub time: Option<String>,
    pub ingredients: Vec<String>,
    pub steps: Vec<String>,
    pub tags: Vec<String>,
    pub source: Option<String>,
    pub image: Option<String>,
    pub extra: HashMap<String, serde_json::Value>,
    pub ingredient_count: usize,
    pub step_count: usize,
}

#[derive(Serialize)]
pub struct RecipeSourceStats {
    pub source_id: String,
    pub source_name: String,
    pub source_icon: String,
    pub recipe_count: usize,
}

#[derive(Serialize)]
pub struct RecipeTypeStats {
    pub name: String,
    pub recipe_count: usize,
}

#[derive(Serialize)]
pub struct RecipeStats {
    pub total_recipes: usize,
    pub by_source: Vec<RecipeSourceStats>,
    pub by_type: Vec<RecipeTypeStats>,
}

#[derive(Serialize)]
pub struct CraftingData {
    pub sources: Vec<RecipeSourceInfo>,
    pub recipes: Vec<RecipeWithComputed>,
    pub stats: RecipeStats,
}
